bazaarvoice/emodb
# Convert role and API management to REST API

## Github Issue #

[50](https://github.com/bazaarvoice/emodb/issues/50)

## What Are We Doing Here?

This unfortunately sizable PR addresses a single purpose: convert role and API key management to a publicly-accessible RESTful front end.  As much as possible this was broken into smaller, logical commits.  The general flow the PR is:

- First several commits:
  - Small fixes or modifications.  These are typically minor things that, on their own, likely wouldn't warrant a PR, but each one blocks a bit of what's necessary to convert to a RESTful API.
- ["Added user access control API and client"](https://github.com/bazaarvoice/emodb/pull/107/commits/a6fa138a86ed17dbfba26abe35f6a6b59202da76)
   - Defines the public-facing user access control (UAC) API and client implementaiton.
- ["Refactored API key and role management as public resources"](https://github.com/bazaarvoice/emodb/pull/107/commits/000711edaa7e452f717956086aae1b0fa95f3c3c)
   - Building on the previous commit, this is where the server backend implements the resources defined by the UAC client.  This and the previous commit represent the bulk of this PR.
- ["Updated documentation"](https://github.com/bazaarvoice/emodb/pull/107/commits/1f08440e98183cb1958d3259824c8488e3cba019)
   - Adds and updates MD files to define the new API.  This may actually be a good starting point to reference what, from a client perspective, has changed with this PR.
- Remaining commits:
   - The remaining commits fall into one of two categories:
      1. Additional features not core to role and API key management but which are desirable.  These were always intended to be part of the implementation, but to keep the core changes separate they were done in separate commits.  For example, the ability to query whether a particular role or key has permission to perform _X_ is useful but not critical, so it that functionality is added here.
       2. Very small updates found after the rest of the work was done.  Not really large enough to warrant fixing and rebasing all of the previous commits.

## How to Test and Verify

1. Check out this PR
2. Go through the API key management README
3. Create roles and API keys to verify all works as expected.  Much of this is already documented in the README so it would be redundant to repeat here.

## Risk

High.  The actual changes aren't particularly complex, but this PR pulls out the old role and API key management system and replaces it with a new one, so any errors may negatively affect the ability to manage these.

Having said that, the backend for roles and API keys is unchanged, so if there were a breaking issue with the interface the system would still work with the existing set of roles and API keys, but until it were fixed there'd be no way to alter that set.

### Level 

`High`

### Required Testing

`Manual`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
